File: src/solvers/hp/hp.cc
//
// This file is part of Gambit
// Copyright (c) 1994-2026, The Gambit Project (https://www.gambit-project.org)
//
// FILE: src/solvers/hp/hp.cc
// Computation of a Nash equilibria using a differentiable homotopy
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//

#include "gambit.h"
#include "solvers/hp/hp.h"
#include "solvers/hp/hpsystem.h"
#include "solvers/logit/path.h"

namespace Gambit::Nash {
std::list<MixedStrategyProfile<double>>
HPStrategySolve(const MixedStrategyProfile<double> &p_prior,
                StrategyCallbackType<double> p_onEquilibrium, HPEventCallbackType p_onEvent,
                const CancelToken &p_cancel)
{
  std::list<MixedStrategyProfile<double>> equilibria;

  HPEquationSystem system(p_prior);
  Vector<double> x = system.ComputeInitialPoint();

  const PathTracer tracer;
  const PathTracer::TraceDirection direction = PathTracer::TraceDirection::Positive;
  const size_t tracking_index = 1; // Track the first variable (t) for orientation
  auto termination_condition = [](const Vector<double> &point) { return point[1] >= 1.5; };
  auto criterion_function = [](const Vector<double> &point,
                               const Vector<double> &tangent) -> double { return point[1] - 1.0; };

  tracer.TracePath(
      [&system](const Vector<double> &point, Vector<double> &lhs) { system.GetValue(point, lhs); },
      [&system](const Vector<double> &point, Matrix<double> &jac) {
        system.GetJacobian(point, jac);
      },
      x, direction, tracking_index, termination_condition,
      [&system, &p_onEvent](const Vector<double> &point) {
        const MixedStrategyProfile<double> profile = system.ExtractEquilibrium(point);
        p_onEvent(HPStepEvent{.profile = profile, .t = point[1]});
      },
      criterion_function, NullCriterionBracketFunction, p_cancel);

  const MixedStrategyProfile<double> equilibrium = system.ExtractEquilibrium(x);
  p_onEquilibrium(equilibrium);
  equilibria.push_back(equilibrium);
  return equilibria;
}
} // namespace Gambit::Nash

File: src/pygambit/cli/hp.py
#
# This file is part of Gambit
# Copyright (c) 1994-2026, The Gambit Project (https://www.gambit-project.org)
#
# FILE: src/pygambit/cli/hp.py
# Command-line driver program for Nash equilibrium computation via the
# Herings-Peeters (2001) homotopy method
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
"""Command-line driver program for Nash equilibrium computation via the
Herings-Peeters (2001) homotopy method.
"""

from __future__ import annotations

import click

import pygambit as gbt

from .common import (
    handle_errors,
    load_game,
    render_profile_csv,
    resolve_strategy_starts,
    version_option,
)

DESCRIPTION = "Compute a Nash equilibrium using the Herings-Peeters (2001) homotopy method"
PROG_NAME = "gambit-hp"


@click.command(
    context_settings={"help_option_names": ["-h", "--help"]},
    help=(
        f"{DESCRIPTION}.\n\n"
        "Reads a game from FILE, or from standard input if FILE is not specified."
    ),
)
@click.argument("file", required=False, default=None)
@click.option(
    "-d",
    "decimals",
    default=6,
    show_default=True,
    type=int,
    help="show equilibria as floating point with DECIMALS digits",
)
@click.option(
    "-n",
    "n_priors",
    type=int,
    default=None,
    help="number of prior distributions to generate randomly (mutually exclusive with -s)",
)
@click.option(
    "-R",
    "seed",
    type=int,
    default=None,
    help="seed the random number generator used to generate prior distributions "
    "(default is to seed from system entropy); requires -n",
)
@click.option(
    "-s",
    "start_file",
    type=str,
    default=None,
    help="file containing prior distributions (mutually exclusive with -n)",
)
@click.option("-q", "--quiet", is_flag=True, help="quiet mode (suppresses banner)")
@click.option(
    "-V",
    "--verbose",
    is_flag=True,
    help="verbose mode (shows each point traced along the homotopy path)",
)
@version_option(DESCRIPTION)
@handle_errors
def main(
    file: str | None,
    decimals: int,
    n_priors: int | None,
    seed: int | None,
    start_file: str | None,
    quiet: bool,
    verbose: bool,
) -> None:
    game = load_game(quiet, DESCRIPTION, file, PROG_NAME)
    priors = resolve_strategy_starts(game, n_priors, seed, start_file)

    def render_event(event: gbt.HPStepEvent) -> None:
        if verbose:
            click.echo(render_profile_csv(event.profile, f"{event.t:.6g}", decimals))

    for prior in priors:
        prior = prior.as_float()
        if verbose:
            click.echo(render_profile_csv(prior, "prior", decimals))
        result = gbt.nash.hp_solve(prior, event_callback=render_event)
        for eq in result.equilibria:
            click.echo(render_profile_csv(eq, "NE", decimals))


if __name__ == "__main__":
    main()
